<a href="https://colab.research.google.com/github/supai-red/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features_AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

Substantive lecture begins 16:00 

1/14/2020

In [0]:
import pandas as pd

In [0]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

### Specific Columns hold specific kinds of data

In [0]:
# Pandas Display Options

pd.set_option('display.max_rows', 15)
pd.set_option('display.max_columns', 15)
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Length: 81, dtype: object

Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [0]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!


Some columns hold float values like the `LotFrontage` column.

In [0]:
# Look at the first ten rows of the `BedroomAbvGr` column.
df['BedroomAbvGr'].head()

0    3
1    3
2    3
3    3
4    4
Name: BedroomAbvGr, dtype: int64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [0]:
df['LotFrontage'].head()

0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
Name: LotFrontage, dtype: float64

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [0]:
# NaN is the most common value in this column. What is a NaN
# NaNs are automatically converted to Floats.
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [0]:
import numpy as np

# What is the datatype of NaN?

type(np.NaN)

float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [0]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets

small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [0]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column

small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [0]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['PricePerSqFt'] = (small_df['SalePrice'] / small_df['TotalSquareFootage'])

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly? The quality of the square footage is higher, or it is more desirable in some way.

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture? The house is less valuable or desirable on a square footage basis.



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [0]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-01-28 00:18:32--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [           <=>      ]  21.72M   771KB/s    in 29s     

2020-01-28 00:19:02 (754 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22776791]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [0]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [0]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [0]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

Numbers are read as strings; 36 months should just be 36; n/a is not a recognized format; positions; dates are strings;...

In [0]:
# Read in the CSV

df = pd.read_csv('LoanStats_2018Q4.csv', header=1) 
# Without header=1 does not work

# Alternatively: df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)

print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128399, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  ...  debt_settlement_flag debt_settlement_flag_date  \
0   14.47%  ...                     Y                  Oct-2019   
1   10.33%  ...                     N                       NaN   
2   15.02%  ...                     N                       NaN   
3   12.98%  ...                     N                       NaN   
4   16.14%  ...                     N                       NaN   

  settlement_status settlement_date settlement_amount settlement_percentage  \
0            ACTIVE        Oct-2019            3479.0                 60.01   
1               NaN             NaN               NaN                   NaN   
2               NaN             NaN               NaN                   NaN   
3               NaN             NaN               NaN                   NaN   
4               NaN             NaN               NaN                   NaN   

   settlement_term  
0             24.0  
1              NaN  
2              NaN  
3              NaN  
4              NaN  

[5 rows x 144 columns]

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [0]:
# We can fix the header problem by using the 'skiprows' parameter

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [0]:
# Sum null values by column and sort from least to greatest
df.isnull().sum().sort_values()

# Almost every column has two missing values.
# last two rows have been read in as columns

inq_fi                                             2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
                                               ...  
orig_projected_additional_accrued_interest    128193
id                                            128397
url                                           128399
member_id                                     128399
desc                                          128399
Length: 144, dtype: int64

In [0]:
# Address the extra NaNs in each column by skipping the footer as well.
df = pd.read_csv('LoanStats_2018Q4.csv', 
                 skiprows=1, 
                 skipfooter=2, 
                 engine='python') # eliminates warning

print(df.shape)
df.head()

# This is one code block that takes care of headers and footers.

(128397, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       6000         6000           6000.0   36 months   
1 NaN        NaN      24000        24000          24000.0   36 months   
2 NaN        NaN      30000        30000          30000.0   36 months   
3 NaN        NaN       8000         8000           8000.0   36 months   
4 NaN        NaN      22400        22400          22400.0   60 months   

  int_rate  ...  debt_settlement_flag debt_settlement_flag_date  \
0   14.47%  ...                     Y                  Oct-2019   
1   10.33%  ...                     N                       NaN   
2   15.02%  ...                     N                       NaN   
3   12.98%  ...                     N                       NaN   
4   16.14%  ...                     N                       NaN   

  settlement_status settlement_date settlement_amount settlement_percentage  \
0            ACTIVE        Oct-2019            3479.0                 60.01   
1               NaN             NaN               NaN                   NaN   
2               NaN             NaN               NaN                   NaN   
3               NaN             NaN               NaN                   NaN   
4               NaN             NaN               NaN                   NaN   

   settlement_term  
0             24.0  
1              NaN  
2              NaN  
3              NaN  
4              NaN  

[5 rows x 144 columns]

In [0]:
df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
                                               ...  
orig_projected_additional_accrued_interest    128191
url                                           128397
member_id                                     128397
desc                                          128397
id                                            128397
Length: 144, dtype: int64

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [0]:
df.shape

# Many columns have no information (where number of nulls = number of rows)

(128397, 144)

In [0]:
df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

print(df.shape)
df.head()

(128397, 140)


loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade  ... debt_settlement_flag debt_settlement_flag_date settlement_status  \
0     C  ...                    Y                  Oct-2019            ACTIVE   
1     B  ...                    N                       NaN               NaN   
2     C  ...                    N                       NaN               NaN   
3     B  ...                    N                       NaN               NaN   
4     C  ...                    N                       NaN               NaN   

  settlement_date  settlement_amount settlement_percentage settlement_term  
0        Oct-2019             3479.0                 60.01            24.0  
1             NaN                NaN                   NaN             NaN  
2             NaN                NaN                   NaN             NaN  
3             NaN                NaN                   NaN             NaN  
4             NaN                NaN                   NaN             NaN  

[5 rows x 140 columns]

In [0]:
df.isnull().sum().sort_values()

loan_amnt                                          0
mo_sin_rcnt_rev_tl_op                              0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
                                               ...  
hardship_status                               128164
hardship_reason                               128164
hardship_dpd                                  128164
hardship_loan_status                          128164
orig_projected_additional_accrued_interest    128191
Length: 140, dtype: int64

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [0]:
# Look at the first 10 values of the int_rate column
df['int_rate'][:10]

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

In [0]:
# Look at a specific value from the int_rate column
df['int_rate'][0]

' 14.47%'

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [0]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

In [0]:
int_rate = ' 14.17%'

In [0]:
int_rate.strip()

'14.17%'

In [0]:
int_rate.strip('%')

' 14.17'

In [0]:
int_rate.strip().strip('%')

'14.17'

In [0]:
type(int_rate.strip().strip('%'))

str

In [0]:
# "Cast" the string value to a float
# Casting is whenever you change one datatype to another datatype
float(int_rate.strip().strip('%'))

14.17

In [0]:
type(float(int_rate.strip().strip('%')))

float

### Write a function to make our solution reusable!

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [0]:
# Test out our function by calling it on our example
int_rate_to_float(int_rate)

14.17

In [0]:
# is the data type correct?
type(int_rate_to_float(int_rate))

float

### Apply our solution to every cell in a column

In [0]:
df['int_rate_float'] = df['int_rate'].apply(int_rate_to_float)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade  ... debt_settlement_flag_date settlement_status settlement_date  \
0     C  ...                  Oct-2019            ACTIVE        Oct-2019   
1     B  ...                       NaN               NaN             NaN   
2     C  ...                       NaN               NaN             NaN   
3     B  ...                       NaN               NaN             NaN   
4     C  ...                       NaN               NaN             NaN   

  settlement_amount  settlement_percentage settlement_term int_rate_float  
0            3479.0                  60.01            24.0          14.47  
1               NaN                    NaN             NaN          10.33  
2               NaN                    NaN             NaN          15.02  
3               NaN                    NaN             NaN          12.98  
4               NaN                    NaN             NaN          16.14  

[5 rows x 141 columns]

In [0]:
# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns
df.dtypes[-5:]

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
int_rate_float           float64
dtype: object

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [0]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False, ascending=False)[:20]

NaN                 20942
Teacher              2090
Manager              1773
Registered Nurse      951
Driver                924
                    ...  
Nurse                 326
Engineer              325
Sales Manager         304
manager               301
Supervisor            270
Name: emp_title, Length: 20, dtype: int64

In [0]:
# How many different unique employment titles are there currently?
len(df['emp_title'].value_counts())

43889

In [0]:
len(df['emp_title'].unique()) # includes NaN

43890

In [0]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20942

What are some possible reasons as to why a person's employment title may have not been provided?

In [0]:
# Create some examples that represent the cases that we want to clean up

examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [0]:
# Write a function to clean up these use cases and increase uniformity.

def clean_title(title):
  if isinstance(title, str):  # Must specify str because of NaNs.
    return title.strip().title()
  else:
    return "Unknown"

for example in examples:
  print(clean_title(example))

Owner
Supervisor
Project Manager
Unknown


In [0]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

[clean_title(example) for example in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [0]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_title)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade  ... debt_settlement_flag_date settlement_status settlement_date  \
0     C  ...                  Oct-2019            ACTIVE        Oct-2019   
1     B  ...                       NaN               NaN             NaN   
2     C  ...                       NaN               NaN             NaN   
3     B  ...                       NaN               NaN             NaN   
4     C  ...                       NaN               NaN             NaN   

  settlement_amount  settlement_percentage settlement_term int_rate_float  
0            3479.0                  60.01            24.0          14.47  
1               NaN                    NaN             NaN          10.33  
2               NaN                    NaN             NaN          15.02  
3               NaN                    NaN             NaN          12.98  
4               NaN                    NaN             NaN          16.14  

[5 rows x 141 columns]

We can use the same code as we did earlier to see how much progress was made.


In [0]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(dropna=False, ascending=False)[:20]

Unknown             20942
Teacher              2557
Manager              2395
Registered Nurse     1417
Driver               1258
                    ...  
Sales Manager         500
Police Officer        440
Nurse                 424
Technician            420
Engineer              412
Name: emp_title, Length: 20, dtype: int64

In [0]:
# How many different unique employment titles are there currently?
len(df['emp_title'].value_counts())

# Eliminated almost 10,000 job titles

34899

In [0]:
# How often is the employment_title null (NaN)?
df['emp_title'].isnull().sum()

0

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [0]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [0]:
df['issue_d'][:10]

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
5    Dec-2018
6    Dec-2018
7    Dec-2018
8    Dec-2018
9    Dec-2018
Name: issue_d, dtype: object

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [0]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [0]:
df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [0]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [0]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128392    2018
128393    2018
128394    2018
128395    2018
128396    2018
Name: issue_d, Length: 128397, dtype: int64

Now the month.

In [0]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128392    10
128393    10
128394    10
128395    10
128396    10
Name: issue_d, Length: 128397, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [0]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade  ... settlement_date settlement_amount settlement_percentage  \
0     C  ...        Oct-2019            3479.0                 60.01   
1     B  ...             NaN               NaN                   NaN   
2     C  ...             NaN               NaN                   NaN   
3     B  ...             NaN               NaN                   NaN   
4     C  ...             NaN               NaN                   NaN   

  settlement_term  int_rate_float issue_year issue_month  
0            24.0           14.47       2018          12  
1             NaN           10.33       2018          12  
2             NaN           15.02       2018          12  
3             NaN           12.98       2018          12  
4             NaN           16.14       2018          12  

[5 rows x 143 columns]

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [0]:
df['earliest_cr_line'].head()

0    Jun-2006
1    Oct-1996
2    May-2007
3    Mar-2009
4    Dec-2004
Name: earliest_cr_line, dtype: object

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], 
                                        infer_datetime_format=True)

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [0]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

df['days_from_earliest_credit_to_issue'].head()

0    4566
1    8096
2    4232
3    3562
4    5113
Name: days_from_earliest_credit_to_issue, dtype: int64

What's oldest credit history that was involved in Q4 2018? 

In [0]:
df['days_from_earliest_credit_to_issue'].describe()

count    128397.000000
mean       5859.918464
std        2886.617678
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.